### PR TITLE
ZA | ITP-May-2025 | Revive Munashe Mapfumo | Module-Data-Groups | Sprint-2

### PR DESCRIPTION
**Pull request by rivakutu**
## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with REGION | COHORT_NAME | FIRST_NAME LAST_NAME | PROJ_NAME
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [x] My changes meet the [requirements](./README.md) of this task

## Changelist

Fixed incorrect data access for address-related entries by correcting object references.

Refactored object manipulation logic using dot notation for clarity and consistency.

Simplified recipe logging and enhanced ingredient handling for better output formatting.

Improved query string parsing to correctly handle edge cases and malformed input.

Implemented tally function to count occurrences of each item, with thorough test coverage.

Added comprehensive test cases for tally, invert, lookup, and contains functions.

Enhanced error handling in test suites to gracefully process invalid input.

Validated input types across functions to ensure safe and predictable behavior.

Streamlined object swapping and inversion logic, including input validation and export/import refinements.

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: Sprint-2/implement/contains.js [hunk 1]
```diff
@@ -1,3 +1,10 @@
-function contains() {}
+function contains(obj, propertyName) {
+  // Validating if input is an object
+  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
+    throw new TypeError("Input must be an object");
+  }
 
+  // checking if property exists inside the object
+  return propertyName in obj;
```
Comment (cjyuan) [inline Sprint-2/implement/contains.js:8]: Consider the following two approaches for determining if an object contains a property:
```
  let obj = {}, propertyName = "toString";
  console.log( propertyName in obj );                // true
  console.log( Object.hasOwn(obj, propertyName) );   // false
```
Which of these approaches suits your needs better?
For more info, you can look up `JS "in" operator vs Object.hasOwn`.

File: Sprint-2/stretch/mode.test.js [hunk 2]
```diff
@@ -13,20 +13,20 @@ const calculateMode = require("./mode.js");
...
+    expect(calculateMode([3, 4, 1, 5, 3, 1, 1])).toEqual(1);
   });
 
   test("returns the first mode in case of multiple modes", () => {
-    const nums = [1, 2, 2, 3, 3];
-
-    expect(calculateMode(nums)).toEqual(2);
+    expect(calculateMode([1, 2, 2, 3, 3])).toEqual(2);
+    expect(calculateMode([1, 1, 2, 7, 7])).toEqual(1);
+    expect(calculateMode([1, 5, 5, 3, 3])).toEqual(5);
```
Comment (cjyuan) [inline Sprint-2/stretch/mode.test.js:24]: Could also include a few cases where the modes interleave. For examples
```
 expect(calculateMode([5, 3, 3, 1, 5])).toEqual(5);   // mode appears as the the first and last number in the array
 expect(calculateMode([9, 2, 4, 5, 3, 3, 4, 1, 1, 2])).toEqual(2); // More than 2 modes
```

File: Sprint-2/stretch/mode.test.js [hunk 3]
```diff
@@ -13,20 +13,20 @@ const calculateMode = require("./mode.js");
...
   });
 
   test("ignores non-number values", () => {
-    const nums = [1, 3, "2", 2, 3, null];
-
-    expect(calculateMode(nums)).toEqual(3);
+    expect(calculateMode([1, 3, "2", 2, 3, null])).toEqual(3);
+    expect(calculateMode([1, 3, "2", undefined, 3, null])).toEqual(3);
+    expect(calculateMode(["2", undefined, 1, null])).toEqual(1);
```
Comment (cjyuan) [inline Sprint-2/stretch/mode.test.js:30]: Could also include an array with no numbers.

**Review by cjyuan — COMMENTED** [review]
Codes are well written! I only have a few suggestions!

File: Sprint-2/debug/recipe.js [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
 // This program should log out the title, how many it serves and the ingredients.
 // Each ingredient should be logged on a new line
 // How can you fix it?
+// by typing recipe.ingredients
 
 const recipe = {
   title: "bruschetta",
   serves: 2,
-  ingredients: ["olive oil", "tomatoes", "salt", "pepper"],
+  ingredients: ["\nolive oil", "tomatoes\n", "salt\n", "pepper"],
```
[the remark this answers is not in the document]
Comment (rivakutu) [reply Sprint-2/debug/recipe.js]: ive left the array unchanged and used .join add a new line after each ingredient

File: Sprint-2/implement/contains.js (hunk 1 above)
Comment (rivakutu) [reply Sprint-2/implement/contains.js]: switched to Object.hasOwn

File: Sprint-2/stretch/mode.test.js (hunk 3 above)
Comment (rivakutu) [reply Sprint-2/stretch/mode.test.js:32]: added a case where array only contains non numbers only
